### ISSUE TITLE
ePubs warn on CSS selector specifies absolute position

### ISSUE THREAD

**Issue by jamtur01** [jgm/skylighting#32]
I am building ePubs with Pandoc (2.1) and I note that:

https://github.com/jgm/skylighting/blob/master/src/Skylighting/Format/HTML.hs#L158
https://github.com/jgm/skylighting/blob/master/src/Skylighting/Format/HTML.hs#L161

Both result in `position: absolute` being set. When validating ePubs built with Pandoc I see:

```
WARNING(CSS-017): /Users/user/src/book/output/TheBook.epub/EPUB/text/ch010.xhtml(32,5): CSS selector specifies absolute position.
WARNING(CSS-017): /Users/user/src/book/output/TheBook.epub/EPUB/text/ch010.xhtml(35,5): CSS selector specifies absolute position.
```
In most readers this doesn't pose a problem but many import/process ePubs and choke on warnings - Google Play Books is an example here. I don't know what the solution is or if there is a solution or if it's worth fixing (and from reading #4128 it seems fairly complex). :) 

c.f. https://github.com/jgm/pandoc/issues/4245

**Comment by dbaynard** [comment]
`position: absolute` is needed for line spacing. There is a workaround; I'll track it down. Meanwhile it should be possible to override this css using an external file; what happens if you supply an external css file with the following?

```css
pre.numberSource a.sourceLine:empty { position: relative; }
pre.numberSource a.sourceLine::before { position: relative; }
```

Also, what media type does the conversion use?

**Comment by jamtur01** [comment]
It doesn't change the error. I presume because skylighting prefixes the CSS into the top of every file so specifying it externally doesn't change that.

The media-type of the resulting HTML files is:

` media-type="application/xhtml+xml" `

**Comment by dbaynard** [comment]
That's odd. Supplying `-c override.css` with pandoc is overriding that behaviour, at least with 2.0.5 — I can't check whether it's producing the error or not, though. I'll check 2.1 but I won't be able to do it until Monday.

I meant media type for media queries (`print` vs `screen`). I suspect `screen`, as long lines of code don't wrap. The workarounds are slightly different for each case. In this case, I'd suggest to override as above and either use

```css
pre.numberSource div.sourceLine::after {
    content: "\2009"; // Thin space: &thinsp;
}
```

or fix the line height, but if you can't override this behaviour (or the errors) I don't know what to suggest. Incidentally, if you can't override the css with the `-c` flag that should be a pandoc bug. Is this a change you've seen since upgrading 2.0.5 → 2.1?

**Comment by dbaynard** [comment]
OK, it looks like skylighting's css is not valid for epub (<http://www.idpf.org/epub/301/spec/epub-contentdocs.html#sec-css-2.1>) which I'll look to fix. I've seen the warnings myself, with epubcheck.

The ipdf link has an exception:

> The absolute value of the position property should be used only with XHTML Content Documents whose rendition:layout property [Publications301] has been set to pre-paginated.

I don't know whether that's possible; it seems undesirable.

Also, I've noticed another needed override:

```css
a.sourceLine:empty { position: relative; }
```

**Comment by jamtur01** [comment]
Yep - for some reason - even with those overrides all in place. I still see:

```html
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <title>Getting help and extending Docker</title>
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
```

**Comment by dbaynard** [comment]
Ah — that's expected behaviour. The overrides won't change the hard-coded skylighting css (hence why it needs fixing here) but they will override behaviour — which makes this sound like a bug in epubcheck (or whichever tool is doing the validation). If it _is_ a bug in epubcheck, it might be particularly tricky to fix.

I'll have a think how to solve it in skylighting. In the meantime, perhaps the output could be run through a filter that replaces the `position: absolute` with `position: relative` in the offending files; my hasty attempt to unzip and filter the entire stream didn't quite work.

**Comment by jamtur01** [comment]
Yeah I already do a bit of that find/replace for some other issues so adding the s/absolute/relative/g for that should be fairly easy too.

**Comment by dbaynard** [comment]
It will probably break the line numbering — I'd be interested in seeing a screenshot, if you get a chance. If not, and the output is acceptable, even better.

**Comment by jamtur01** [comment]
I don't use line numbering.

**Comment by dbaynard** [comment]
@jamtur01 Did your modified epub work?

**Comment by jamtur01** [comment]
It did. Worked fine.

**Comment by dbaynard** [comment]
@jamtur01 

> Yeah I already do a bit of that find/replace for some other issues so adding the s/absolute/relative/g for that should be fairly easy too.

Is there anything in your f/r list that should be fixed in skylighting?

**Comment by jamtur01** [comment]
@dbaynard No - it's mostly the mix of LaTeX and Markdown I use. Like turning my LateX admonitions into HTML divs.

**Comment by jagthedrummer** [comment]
I've just run into this same problem trying to create an epub with pandoc that happens to contain a code block. It seems like maybe there should be some flag that could be passed to pandoc to trigger use of `relative` instead of `absolute`.

**Comment by jgm** [comment]
@dbaynard it seems like we need some fixes to skylighting?
Can you propose something?
It isn't good if the CSS we produce isn't valid for EPUBs.

**Comment by dbaynard** [comment]
Yes, I'll take a look this evening.

**Comment by jagthedrummer** [comment]
@dbaynard It's a validation issue, which I think is preventing some people from being able to import/read the epub because their reader tries to validate it before importing.

**Comment by dbaynard** [comment]
Great, that's helpful to know. I'll be able to test with epubcheck.

**Comment by dbaynard** [comment]
I've got a fix, using `position: relative`, inline-blocks and adjusting line lengths. Just testing for other display issues.

```css
.numberSource a.sourceLine::before {
    position: relative;
    left: -1em;
    display: inline-block;
}

.numberSource a.sourceLine {
    left: -4em;
}
```

Lines in the resulting epub files don't wrap, though. I'm trying to fix that. (It should behave like `@media print` even though it is `@media screen`.)

**Comment by dbaynard** [comment]
> Lines in the resulting epub files don't wrap, though. I'm trying to fix that. (It should behave like `@media print` even though it is `@media screen`.)

I haven't changed this, mainly because it is impossible to support the full behaviour with just skylighting yet it is straightforward to apply custom css in pandoc that should override this behaviour — that said, I would like pandoc to wrap these lines when producing an epub by default. If there's demand, I'll do it more urgently.

**Comment by jgm** [comment]
dbaynard <notifications@github.com> writes:

> I haven't found a good way to identify within a css file whether it is applied to an epub or to standard html. So, in order to ensure code wraps by default it would have to apply to the default html output from pandoc, and thus all html code would wrap — which _would_ be a change to pandoc's behaviour.

But maybe a desirable change?  Should this be raised on pandoc-discuss?

> Given that all that is needed is the code below, it may be better to add it in pandoc itself.
>
> ```css
> @media screen {
>     div.sourceCode {
>         overflow: visible;
>     }
>     code.sourceCode {
>         white-space: pre-wrap;
>     }
> }
> ```

Just to be clear, this should go in data/epub.css?

**Comment by jgm** [comment]
dbaynard <notifications@github.com> writes:

>> But maybe a desirable change?  Should this be raised on pandoc-discuss?
>
> Absolutely, a desirable change. I've just joined the group.
>
> There is one downside, though. Without line numbers, it is not always clear whether code has wrapped. It's possible to wrap numbered lines only, though I was hoping it might be possible to introduce some indicator that a line has wrapped.
>
>> Just to be clear, this should go in data/epub.css?
>
> Yes. I've just tested that, and it overrides the default skylighting behaviour. Shall I PR?

Would the PR still be needed if we just made the change described
above to skylighting (enabling wrapping)?

If yes, then go ahead and make the PR. If no, then we should decide
first whether to make that change.

**Comment by dbaynard** [comment]
Actually, this is more complicated than I'd thought. I think I have to re-do the generated html to include elements to take the line numbers (as in jgm/pandoc#4386). With `position: relative` instead of `position: absolute` it plays havoc with subsequent line spacing. The current output is usable, though not pretty.

To be clear, this only affects pdfs produced via html, but once the changes are made to pandoc's `data/epub.css` it would affect that, too.

![wrapping-problems](https://user-images.githubusercontent.com/870937/39939568-9ae9b100-554e-11e8-9174-141cbc5484ea.png)

I'll fix jgm/pandoc#4386 this weekend. Users who supply their own css may have to change that css.
